### ISSUE TITLE
java.lang.VerifyError: Illegal type at constant pool entry 156 in class 'etc'

### ISSUE THREAD

**Issue by joaomneto** [Kotlin/kotlinx.coroutines#1533]
Hi there, I'm using the following code to read a Databuffer flux in string boot using coroutines and accumulate bytes in an intermediate buffer (because S3 multipart async upload requires 5MB part minimum). Using fold in the flow from created from the flux produces this exception.

This run perfectly when launching the server using `./gradlew bootRun` but breaks when running `java -jar xxx.jar` when the jar was produced via `./gradlew bootJar` because the BootRun task adds `-Xverify:none` and `-XX:TieredStopAtLevel=1` to the jvm args.

```
Exception in thread "sdk-async-response-0-0" kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for CancellableContinuation(DispatchedContinuation[Unconfined, Continuation at com.talkdesk.api.reporting.exploreemail.services.ExternalStorageService.initiateMultipartUpload(ExternalStorageService.kt:147)@4cdabf65]){CreateMultipartUploadResponse(Bucket=td-infra-stg-us-east-1-s3-reportingexplore-email-files, Key=The-A-Team/2019-09-13/ca513e2283a0406ba7bfc950d659c429.csv, UploadId=a93JiolPp5fFcjTkb0AUxf7WmRwqW_xAx.lM47PvPaO.q6vXDuoMifUX8mKpWM1NFx2vmvmh9bpW_hiiki6loL9JRoYlEGLoh3OFuo8G6bLVIjUDuVP5p8lV2HUhyZC_, ServerSideEncryption=AES256)}@6ff72e73. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(Dispatched.kt:278)
	at kotlinx.coroutines.DispatchedKt.resumeUnconfined(Dispatched.kt:415)
	at kotlinx.coroutines.DispatchedKt.dispatch(Dispatched.kt:298)
	at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:250)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:260)
	at kotlinx.coroutines.CancellableContinuationImpl.resumeWith(CancellableContinuationImpl.kt:189)
	at kotlinx.coroutines.future.ContinuationConsumer.accept(Future.kt:178)
	at kotlinx.coroutines.future.ContinuationConsumer.accept(Future.kt:170)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncApiCallTimeoutTrackingStage.lambda$execute$2(AsyncApiCallTimeoutTrackingStage.java:69)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryResponseIfNeeded(AsyncRetryableStage.java:136)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.retryIfNeeded(AsyncRetryableStage.java:120)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.AsyncRetryableStage$RetryExecutor.lambda$execute$0(AsyncRetryableStage.java:107)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeAsyncHttpRequestStage.lambda$executeHttpRequest$1(MakeAsyncHttpRequestStage.java:138)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.VerifyError: Illegal type at constant pool entry 156 in class REDACTED$upload$$inlined$fold$1
Exception Details:
  Location:
    REDACTED$upload$$inlined$fold$1.emit(Ljava/lang/Object;Lkotlin/coroutines/Continuation;)Ljava/lang/Object; @274: invokestatic
  Reason:
    Constant pool index 156 is invalid
  Bytecode:
    ${REDACTED}
  Stackmap Table:
    same_frame(@36)
    append_frame(@46,Object[#48])
    append_frame(@84,Object[#5],Object[#5])
    full_frame(@225,{Object[#2],Object[#5],Object[#79],Object[#48],Object[#5],Object[#5],Object[#48],Object[#5],Integer,Object[#79],Object[#81],Object[#83],Object[#74],Integer,Integer},{})
    same_locals_1_stack_item_frame(@227,Object[#83])
    full_frame(@386,{Object[#2],Object[#5],Object[#79],Object[#48],Object[#5],Object[#5]},{})
    full_frame(@481,{Object[#2],Object[#5],Object[#79],Object[#48],Object[#5],Object[#5],Object[#48],Object[#5],Top,Object[#79],Object[#81],Object[#83],Object[#74],Top,Integer,Top,Object[#83],Object[#207]},{Object[#5]})
    chop_frame(@499,1)
    full_frame(@550,{Object[#2],Object[#5],Object[#79],Object[#48],Object[#5],Object[#5]},{})

	at REDACTED.upload$suspendImpl(ExternalStorageService.kt:174)
	at REDACTED$upload$2.invokeSuspend(ExternalStorageService.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.ResumeModeKt.resumeMode(ResumeMode.kt:67)
	at kotlinx.coroutines.DispatchedKt.resume(Dispatched.kt:319)
	at kotlinx.coroutines.DispatchedKt.resumeUnconfined(Dispatched.kt:49)
	... 29 more
```

Dependencies:
```
id("org.springframework.boot") version "2.2.0.M5"
id("io.spring.dependency-management") version "1.0.8.RELEASE"
kotlin("jvm") version "1.3.50"
kotlin("plugin.spring") version "1.3.41"
implementation("org.springframework.boot:spring-boot-starter-actuator")
implementation("org.springframework.boot:spring-boot-starter-webflux")
implementation("org.springframework.boot:spring-boot-starter-data-jdbc")
implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.1")
implementation("org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:1.3.1")
implementation("org.jetbrains.kotlinx:kotlinx-coroutines-reactor:1.3.1")
implementation("org.jetbrains.kotlinx:kotlinx-coroutines-reactive:1.3.1")
```

``` 
val flow = WebClient
                .builder()
                .baseUrl("some_url")
                .build()
                .get()
                .retrieve()
                .bodyToFlux(DataBuffer::class.java)
                .asFlow()

flow.fold(ByteBuffer.allocate(intermediateBufferLength), { buffer, value -> {
            
///Eventually I call the AWS async client

s3AsyncClient.uploadPart(
        UploadPartRequest.builder()
                .bucket(bucketName)
                .key(key)
                .partNumber(partNumber)
                .uploadId(uploadId)
                .build(),
        asyncRequestBody
).await()

})
```

**Comment by qwwdfsad** [comment]
Thanks for the report!
`VerifyError` is an indicator, that this is a Kotlin compiler bug.
I will minimize and report it to our compiler folks and hopefully will find a temporary workaround for you.

**Comment by joaomneto** [comment]
I'm working with `-Xverify:none` for now. I'll remove it upon a fix on your end. 🙏

**Comment by qwwdfsad** [comment]
Could you please specify what return type has `uploadPart`?

**Comment by joaomneto** [comment]
It's part of S3AsyncClient from AWS:

https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/services/s3/S3AsyncClient.html#uploadPart-java.util.function.Consumer-software.amazon.awssdk.core.async.AsyncRequestBody-

I'm using the `CompletableFuture.await` extension function on it's return.

**Comment by qwwdfsad** [comment]
Thanks! Anyway, have hard times reproducing it locally, is it possible for you to share the reproducing project?

The main suspect here is `flow.fold` inlining (thus the potential workaround is to copy-paste `fold` operator implementation into `fold2` without `inline` modifier) and the outer bytecode shape, but I am unable to guess the bytecode shape here

**Comment by joaomneto** [comment]
I'll try to create a stripped down version, that still generates the problem.

**Comment by qwwdfsad** [comment]
Thanks a lot. Do not spend a lot of effort on stripping though, even the most complicated version of the code is enough as long as it fails with `VerifyError`.

>I'm working with -Xverify:none for now. I'll remove it upon a fix on your end. 🙏

As the bug is in the compiler, ETA of the fix is no earlier than 1.3.60.
I would strongly recommend using a workaround
```
public suspend fun <T, R> Flow<T>.foldNonInlined(
    initial: R, operation: suspend (acc: R, value: T) -> R) = fold(initial, operation)
```

to avoid any problems with `verify:none` in production.

**Comment by joaomneto** [comment]
I'll give it a go. Thanks. 🙏

**Comment by mkulak** [comment]
I hit the same issue. Do you need another reproducer for that one?

**Comment by qwwdfsad** [comment]
@mkulak yes, it would be really nice to have one

**Comment by mkulak** [comment]
@qwwdfsad  
After updating from `1.3.51` to `1.3.61` bug disappeared. But after some code changes it reappeared again. You can find project here: https://github.com/kvarto/kvarto/tree/compiler-error
Just checkout branch `compiler-error` and run `./gradlew test --tests *VertxHttpClientTest --info`
Result: 
```
java.lang.VerifyError: Illegal type at constant pool entry 39 in class io.kvarto.http.client.impl.VertxHttpClient$sendRequest$$inlined$map$1$2
    Exception Details:
      Location:
        io/kvarto/http/client/impl/VertxHttpClient$sendRequest$$inlined$map$1$2.emit(Ljava/lang/Object;Lkotlin/coroutines/Continuation;)Ljava/lang/Object; @44: invokestatic
      Reason:
        Constant pool index 39 is invalid
```
Env:
> Kotlin 1.3.61
java version "1.8.0_102"
Java(TM) SE Runtime Environment (build 1.8.0_102-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.102-b14, mixed mode)
Gradle 5.2.1
Macos Catalina 10.15.2

I didn't create minimized example for this one.   
If I add `-noverify` option test passes successfully.

**Comment by qwwdfsad** [comment]
Thanks for the reproducer!
As a workaround, you can replace ` body.map { Buffer.buffer(it) }` with
```
private fun ByteArray.toBuffer() = Buffer.buffer(this)

// sendRequest body, line 36
body.map { it.toBuffer() }
```
in `sendRequest`

It appears to be a problem during inlining of bytecode complied with different javac version

**Comment by qwwdfsad** [comment]
This is https://youtrack.jetbrains.com/issue/KT-30744 and it will be fixed in 1.3.70.
Verified that bug no longer reproduces with 1.3.70-eap-42

**Comment by vikasrathore** [comment]
@joaomneto please check this i am getting this issue 
[Fatal Exception: java.lang.VerifyError: kotlinx/coroutines/bv$c](https://github.com/Kotlin/kotlinx.coroutines/issues/1894)
